File: algorithms/algorithms-python/leetcode/Question_0016_3Sum_Closest.py
#!/usr/bin/python
# coding: utf-8

class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        res = nums[0] + nums[1] + nums[2]
        diff = abs(res - target)
        nums.sort()
        for n in range(len(nums)):
            l, r = n + 1, len(nums) - 1
            while l < r:
                num = nums[n] + nums[l] + nums[r]
                ndiff = abs(num - target)
                if ndiff < diff:
                    diff = ndiff
                    res = num
                if num < target:
                    l += 1
                else:
                    r -= 1
        return res

File: algorithms/algorithms-cplusplus/leetcode/Question_0016_3Sum_Closest.cc
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
 public:
  int threeSumClosest(vector<int>& nums, int target) {
    int res = nums[0] + nums[1] + nums[2];
    int diff = abs(res - target);
    sort(nums.begin(), nums.end());
    for (size_t k = 0; k < nums.size(); ++k) {
      size_t l = k + 1, r = nums.size() - 1;
      while (l < r) {
        int sum = nums[k] + nums[l] + nums[r];
        int ndiff = abs(sum - target);
        if (ndiff < diff) {
          diff = ndiff;
          res = sum;
        }
        if (sum < target) {
          ++l;
        } else {
          --r;
        }
      }
    }
    return res;
  }
};
